sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala

```diff
@@ -723,6 +724,158 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
       }
     }
   }
+
+  test("SPARK-46272: create table as select") {
+    val cls = classOf[SupportsExternalMetadataDataSource]
+    withTable("test") {
+      sql(
+        s"""
+           |CREATE TABLE test USING ${cls.getName}
+           |AS VALUES (0, 1), (1, 2)
+           |""".stripMargin)
+      checkAnswer(sql("SELECT * FROM test"), Seq(Row(0, 1), Row(1, 2)))
+      sql(
+        s"""
+           |CREATE OR REPLACE TABLE test USING ${cls.getName}
+           |AS VALUES (2, 3), (4, 5)
+           |""".stripMargin)
+      checkAnswer(sql("SELECT * FROM test"), Seq(Row(2, 3), Row(4, 5)))
+      sql(
+        s"""
+           |CREATE TABLE IF NOT EXISTS test USING ${cls.getName}
+           |AS VALUES (3, 4), (4, 5)
+           |""".stripMargin)
+      checkAnswer(sql("SELECT * FROM test"), Seq(Row(2, 3), Row(4, 5)))
+    }
+  }
+
+  test("SPARK-46272: create table as select - error cases") {
+    val cls = classOf[SupportsExternalMetadataDataSource]
+    // CTAS with too many columns
+    withTable("test") {
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(
+            s"""
+               |CREATE TABLE test USING ${cls.getName}
+               |AS VALUES (0, 1, 2), (1, 2, 3)
+               |""".stripMargin)
+        },
+        errorClass = "INSERT_COLUMN_ARITY_MISMATCH.TOO_MANY_DATA_COLUMNS",
```

Comment:
I'm a bit confused here. So for this data source, we probably don't check anything when creating the table. But when load the table, and get the `Table` instance in `V2SessionCatalog`, shall we check if `Table#schema` match what we stored in HMS?

Reply:
This is actually tricky, and behavior depends on the implementation of the data source.

In `createTable`, we store the query's schema (col1, col2, col3) in HMS, but this data source's `getTable` method does not take into account the input `schema`.

So, when the DSv2 relation is being created, it uses `table.columns.asSchema` which is the default implementation of the schema (i, j).  
https://github.com/apache/spark/blob/bacdb3b5fec9783f46042764eeee80eb2a0f5702/sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala#L210-L211

However, if the data source utilize user-defined schema in `getTable`, it won't throw this exception.

Reply:
I think we should enforce something here. When calling `TableProvider.getTable` with user-specified schema, we should make sure the returned table reports the same schema as the user-specified schema (probably ignore nullability). It's actually a documented requirement
```
   * Return a {@link Table} instance with the specified table schema, partitioning and properties
   * to do read/write. The returned table should report the same schema and partitioning with the
   * specified ones, or Spark may fail the operation.
```